app/services/ConfluenceService.scala

```diff
@@ -0,0 +1,227 @@
+import utils.FutureExt
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+
+case class ConfluenceSynchroContext(project: Project, confluenceProjectParentPage: ConfluencePage, directory: Directory)
+
+class ConfluenceService @Inject()(confluenceClient: ConfluenceClient,
+                                  projectRepository: ProjectRepository, branchRepository: BranchRepository, directoryService: DirectoryService
+                                 )(implicit ec: ExecutionContext) extends Logging {
+
+  def refreshProjectInConfluence(projectId: String): Future[Either[ConfluenceError, Seq[Seq[ConfluencePage]]]] = {
+    projectRepository.findById(projectId) match {
+      case None => Future.successful(Left(ConfluenceError(s"Project $projectId not found", ConfluenceError.PROJECT_NOT_FOUND)))
+      case Some(project) => {
+        branchRepository.findByProjectIdAndName(projectId, project.stableBranch) match {
+          case None => Future.successful(Left(ConfluenceError(s"Branch $projectId ${project.stableBranch} not found", ConfluenceError.STABLE_BRANCH_NOT_FOUND)))
+          case Some(branch) => {
+            if (project.confluenceParentPageId.isEmpty) {
+              Future.successful(Left(ConfluenceError(s"Project $projectId has no confluence page id", ConfluenceError.PROJECT_NOT_SETUP)))
+            } else {
+              directoryService.buildDirectoryTree(branch.id) match {
+                case None => Future.successful(Left(ConfluenceError(s"No directories for $projectId ${project.stableBranch}", ConfluenceError.ROOT_DIRECTORY_NOT_FOUND)))
+                case Some(directoriesWithPagesAsTree) => {
+                  directoryService.walkOnDirectoryTree[ConfluenceError, Seq[ConfluencePage]](directoriesWithPagesAsTree,
+                    (d: Directory) => refreshDirectoryInConfluence(project, project.confluenceParentPageId.get.toLong, d),
+                    (t: Throwable) => ConfluenceError(t.getMessage))
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  def refreshDirectoryInConfluence(project: Project, confluenceProjectParentPageId: Long, directory: Directory): Future[Either[ConfluenceError, Seq[ConfluencePage]]] = {
+    confluenceClient.getConfluencePage(confluenceProjectParentPageId).map {
+      case Right(confluenceProjectParentPage) => {
+        if (directory.isRoot()) {
+          refreshRootDirectoryInConfluence(project, confluenceProjectParentPageId, directory)
+        } else {
+          refreshInternalDirectoryInConfluence(confluenceProjectParentPage.space.key, directory)
+        }
+      }
+      case Left(error) => Future.successful(Left(error))
+    }.flatten
+  }
+
+  private def refreshRootDirectoryInConfluence(project: Project, confluenceProjectParentPageId: Long, rootDirectory: Directory): Future[Either[ConfluenceError, Seq[ConfluencePage]]] = {
```

Comment:
Please add line returns to aerate this code between the `case` And maybe some comments if it useful to quickly understand what the code does bellow.

Would it be possible also to simply and have less imbricated pattern match? The cyclomatic complexity is quite high.

Reply:
Maybe a simple `Future` instead of a `Future[Either[ConfluenceError` could simplify the matching and the error recovery?